### PR TITLE
feat(sns): add the MaximumMessageSize topic attribute for 1 MiB payloads

### PR DESCRIPTION
**Pull request by slang25**
## Summary

AWS [raised the SNS payload ceiling to 1 MiB](https://aws.amazon.com/about-aws/whats-new/2026/09/amazon-sns-1mib-support/) behind a new per-topic `MaximumMessageSize` attribute, so this wires it up: range validation on `CreateTopic`/`SetTopicAttributes`, per-topic sizing in `Publish`/`PublishBatch`, the subscription constraints a topic above 256 KiB has to meet, and `AWS::SNS::Topic` passthrough. The default stays `262144`, so a topic that doesn't set the attribute behaves exactly as it did before.

I checked all of this against the real API with the CLI rather than guessing, and every error code and message in here came back off the wire — including two things I wouldn't have assumed: `GetTopicAttributes` omits the attribute entirely until you set it, and the 100-subscription cap is only enforced on `SetTopicAttributes`, not on `Subscribe` (so a raised topic can drift past 100 and only trips up the next time you set the attribute).

Two drive-by fixes in the same code path, also checked against AWS: `Subject` was being counted towards the publish limit when AWS doesn't count it, and `PublishBatch` was returning SQS's `Batch requests cannot be longer than N bytes.` wording rather than SNS's.

## Type of change

- [ ] Bug fix (`fix:`)
- [x] New feature (`feat:`)
- [ ] Breaking change (`feat!:` or `fix!:`)
- [ ] Docs / chore

## AWS Compatibility

Verified against live SNS in `us-east-1` with `aws-cli/2.36.49`, against the model in aws/aws-sdk-net@5d96b4b. Probe topics and queues have been cleaned up.

The two fixes above were both wrong before: a publish of exactly `262144` bytes with a `Subject` was rejected when AWS accepts it, and the `PublishBatch` error text was SQS's rather than SNS's (`The length of all the messages put together is more than the limit.`).

## Checklist

- [x] `./mvnw test` passes locally — 25170 tests, 0 failures. Three Docker-backed tests error on my machine (`SecurityGroupFirewallDockerIntegrationTest`, `EksImdsDockerIntegrationTest`, `RdsAwsIntegrationTest`), all network-namespace/container things unrelated to this change.
- [x] New or updated integration test added
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)

### CODE REVIEW

File: src/main/java/io/github/hectorvent/floci/services/sns/SnsService.java [hunk 1]
```diff
@@ -269,10 +282,97 @@ public void setTopicAttributes(String topicArn, String attributeName,
...
+
+    private static int maxMessageSize(Topic topic) {
+        String value = topic.getAttributes().get(MAXIMUM_MESSAGE_SIZE);
+        if (value == null) {
+            return DEFAULT_MAX_MESSAGE_SIZE;
+        }
+        try {
+            return Integer.parseInt(value);
+        } catch (NumberFormatException ignored) {
+            return DEFAULT_MAX_MESSAGE_SIZE;
```
[the remark this answers is not in the document]
Comment (slang25) [reply src/main/java/io/github/hectorvent/floci/services/sns/SnsService.java]: Good spot, agreed — fixed in b4eac7e.

`maxMessageSize` now treats a parseable but out-of-range persisted value exactly like a nonnumeric one and falls back to `262144`. So a legacy topic carrying, say, `5242880` can't wave a publish past the AWS ceiling, and one carrying `0` or a negative value can't reject every publish or silently gate `Subscribe`. Unit tests cover both the publish path and the protocol check.

File: src/main/java/io/github/hectorvent/floci/services/sns/SnsService.java [hunk 2]
```diff
@@ -1810,12 +1914,14 @@ private String sqsArnToUrl(String arn) {
         return AwsArnUtils.arnToQueueUrl(arn, baseUrl);
     }
 
-    private static int computePublishSize(int messageBytes, String subject,
+    /**
+     * Payload size as AWS counts it against {@code MaximumMessageSize}: UTF-8 body bytes plus,
+     * per attribute, its name, data type and value. {@code Subject} is excluded -- a publish of
+     * exactly the limit succeeds however long its subject is.
```
[the remark this answers is not in the document]
Comment (slang25) [reply src/main/java/io/github/hectorvent/floci/services/sns/SnsService.java:1932]: You're right that we don't validate `Subject` anywhere, but I don't think it belongs in this PR.

Before this change a subject was only ever rejected if it was large enough on its own to push the payload over 256 KiB — nothing AWS actually rejects (over 100 characters, line breaks, control characters) was being caught either way. So it's a pre-existing gap rather than one this change opens up.

The rules and the exact error wording want checking against the live API like everything else here, and guessing at them risks rejecting publishes that real SNS is happy with. I'd rather do it as its own PR — I'll open an issue for it 🙂

File: src/test/java/io/github/hectorvent/floci/services/sns/SnsMaximumMessageSizeTest.java [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * The {@code MaximumMessageSize} topic attribute, which AWS added in September 2026 to raise the
+ * SNS payload ceiling from 256 KiB to 1 MiB. Every expectation here -- error codes, error wording,
+ * boundary values and the subscription constraints -- was taken from the live API.
+ */
+class SnsMaximumMessageSizeTest {
```
[the remark this answers is not in the document]
Comment (slang25) [reply src/test/java/io/github/hectorvent/floci/services/sns/SnsMaximumMessageSizeTest.java:24]: Agreed, added in b4eac7e.

`SnsIntegrationTest` now covers `MaximumMessageSize` end to end on both protocols: create/set/get, the attribute being omitted until it's set, an out-of-range value rejected on `CreateTopic` (Query) and `SetTopicAttributes` (JSON 1.1), a 300 KiB publish clearing a raised topic, and the oversized `Publish` and `PublishBatch` errors — asserting codes and messages off the wire rather than off the service.

File: src/main/java/io/github/hectorvent/floci/services/sns/SnsService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (slang25) [reply src/main/java/io/github/hectorvent/floci/services/sns/SnsService.java:1932]: Opened #3959 for it 👍

**Review by pgermosen — APPROVED** [review]
Nice work on this. I checked the range, the default, the 100-subscription and protocol restrictions, and the InvalidParameter / BatchRequestTooLong codes against the SNS API reference and they line up. Good call on the persisted-value fallback and on keeping Subject validation as its own follow-up in #3959. Looks good to me.
